main.cpp

```diff
@@ -5913,6 +6090,738 @@ bool link_command::execute(device_map &devices) {
+            };
+            do_lfs_op(lfs_op);
+            break;
+        }
+
+        case fs_fatfs: {
+            fatfs_op_fn fatfs_op = [&](FATFS *fatfs) {
+                fatfs_ls(fatfs, dir.c_str(), settings.bdev.recursive);
+            };
+            do_fatfs_op(fatfs_op);
+            break;
+        }
+
+        default:
+            fail(ERROR_ARGS, "Unknown filesystem specified");
+    }
+    return false;
+}
+
+bool bdev_mkdir_command::execute(device_map &devices) {
+    auto con = get_single_bootsel_device_connection(devices);
+    setup_bdevfs(con);
+
+    switch (settings.bdev.fs) {
+        case fs_littlefs: {
+            lfs_op_fn lfs_op = [&](lfs_t *lfs) {
+                int err = lfs_mkdir(lfs, settings.filenames[0].c_str());
+                if (err == LFS_ERR_EXIST) {
+                    fos << "Directory already exists\n";
+                } else if (err) {
+                    fos << "LittleFS Error " << err << "\n";
+                }
+            };
+            do_lfs_op(lfs_op);
+            break;
+        }
+
+        case fs_fatfs: {
+            fatfs_op_fn fatfs_op = [&](FATFS *fatfs) {
+                int err = f_mkdir(fatfs, settings.filenames[0].c_str());
+                if (err == FR_EXIST) {
+                    fos << "Directory already exists\n";
+                } else if (err) {
+                    fos << "FatFS Error " << err << "\n";
+                }
+            };
+            do_fatfs_op(fatfs_op);
+            break;
+        }
+
+        default:
+            fail(ERROR_ARGS, "Unknown filesystem specified");
+    }
+
+    fos << "Created directory " << settings.filenames[0] << " on device\n";
+
+    return false;
+}
+
+bool bdev_cp_command::execute(device_map &devices) {
+    auto con = get_single_bootsel_device_connection(devices);
+    setup_bdevfs(con);
+
+    if ((char)(settings.filenames[1].back()) == '/') {
+        int filenamestart = std::max(settings.filenames[0].find_last_of("/") + 1, settings.filenames[0].find_last_of(":") + 1);
+        int filenamelen = settings.filenames[0].length() - filenamestart;
+        settings.filenames[1] += settings.filenames[0].substr(filenamestart, filenamelen);
+    }
+
+    bool srcRemote = false;
+    bool destRemote = false;
+    if ((char)(settings.filenames[0].front()) == ':') {
+        srcRemote = true;
+        settings.filenames[0].erase(0, 1);
+    }
+    if ((char)(settings.filenames[1].front()) == ':') {
+        destRemote = true;
+        settings.filenames[1].erase(0, 1);
+    }
+
+    fos << "Copying " << (srcRemote ? "device " : "local ") << settings.filenames[0] << " to " << (destRemote ? "device " : "local ") << settings.filenames[1] << "\n";
+
+    std::vector<char> data_buf;
+
+    // Read local file
+    if (!srcRemote) {
+        auto infile = get_file_idx(ios::in|ios::binary|ios::ate, 0);
+        auto size = infile->tellg();
+        infile->seekg(0, std::ios::beg);
+        data_buf.resize(size);
```

Comment:
If you accidentally point this at the wrong local filename (e.g. a huge disk-image file), is there a risk that this might cause an out-of-memory error?

Reply:
Yes - should probably check the input file fits in the largest flash size (32MB currently)

Reply:
That's up to the filesystem - both LittleFS and FatFS should throw an error if there isn't enough free space on the device, but I might add a backstop to prevent writing past the end of the block device too